src/Sparrow/Utils/ZstdStream.cs

```
@@ -0,0 +1,361 @@
+                throw new InvalidOperationException("Should never happen, the buffer is full of data that produces not output");
+
+            fixed (byte* pTempBuf = _tempBuffer, pCurVal = _decompressionInput.Span)
+            {
+                if (pTempBuf == pCurVal)
+                    return;
+                Buffer.MemoryCopy(pCurVal, pTempBuf, _tempBuffer.Length, _decompressionInput.Length);
+                _decompressionInput = new Memory<byte>(_tempBuffer, 0, _decompressionInput.Length);
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return Read(new Span<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (_readSemaphore.Wait(10_000) == false)
+                throw new InvalidOperationException("Couldn't acquire write lock for 10 seconds!");
+
+            try
+            {
+                if (_disposeOnce.Disposed)
+                    throw new ObjectDisposedException("Object was already disposed!");
+                while (true)
+                {
+                    int read = DecompressStep(buffer);
+                    if (read != 0)
+                        return read;
+
+                    ShiftBufferData();
+
+                    read = _inner.Read(_tempBuffer, _decompressionInput.Length, _tempBuffer.Length - _decompressionInput.Length);
+                    if (read == 0)
+                        return 0; // nothing left to read
+
+                    _decompressionInput = new Memory<byte>(_tempBuffer, 0, _decompressionInput.Length + read);
+                    read = DecompressStep(buffer);
+                    if (read != 0)
+                        return read;
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // object was already disposed, so nothing to do...
+            }
```

Why are you hiding the exception here?